src/Development/IDE/Core/Compile.hs

```diff
@@ -398,7 +397,19 @@ parseFileContents customPreprocessor dflags filename contents = do
                (Map.fromListWith (++) $ annotations pst,
                  Map.fromList ((noSrcSpan,comment_q pst)
                                   :annotations_comments pst))
-             (warns, errs) = getMessages pst dflags
+             (warns, errs) =
+                let (ws, es) = getMessages pst dflags
+                    (ws', es') = flip Bag.partitionBagWith ws $ \w ->
+                        -- This ensures -Werror=foo is respected
+                        case errMsgReason w of
+                            Reason flag
+                                | flag `EnumSet.member` fatalWarningFlags dflags
```

Comment:
do you understand why we have to do this separately for the parser but not the typechecker?

Reply:
AFAICT, the typechecker uses [`printOrThrowWarnings`](https://github.com/digital-asset/ghc/blob/9dbc5d1e7c5bb82fb122a21df6e7606d71a3e1af/compiler/main/HscTypes.hs#L328-L345), but it isn't used from the parser. In turn, the [`ParserFlags`](https://hackage.haskell.org/package/ghc-8.8.1/docs/src/Lexer.html#ParserFlags) datatype used as part of the `ParserState` includes a field for warning flags, but not for _fatal_ warning flags, so that makes it a bit hard to use `printOrThrowWarnings` from the parser.

We did see that GHC seems to do the right thing, but I'm not sure what they're doing differently.

Reply:
alright this looks fairly reasonable

Reply:
thanks! I think we could extract the common parts out of `printOrThrowWarnings`, while adding the fatal warning flags to `ParserFlags`, but that would require an additional PR to (the fork of) `ghc`, so for now I skipped it.